packages/ag-charts-enterprise/src/features/annotations/callout/calloutScene.ts

```diff
@@ -1,23 +1,360 @@
+
     constructor() {
         super();
-        this.append([this.label, this.start, this.end]);
+        this.append([this.shape, this.label, this.start, this.end]);
+    }
+
+    override drag(datum: CalloutProperties, target: Coords, context: AnnotationContext) {
+        if (datum.locked) return;
+
+        if (this.activeHandle === 'end') {
+            this.dragHandle(datum, target, context);
+        } else {
+            this.dragAll(datum, target, context);
+        }
+    }
+
+    protected override getLabelCoords(datum: CalloutProperties, bbox: _Scene.BBox, coords: LineCoords): _Util.Vec2 {
+        const {
+            bodyBounds = {
+                x: 0,
+                y: 0,
+                width: 0,
+                height: 0,
+            },
+        } = this.getDimensions(datum, bbox, coords) ?? {};
+
+        return {
+            x: bodyBounds.x + (datum.padding ?? 10),
+            y: bodyBounds.y - bodyBounds.height / 2,
+        };
+    }
+
+    protected override getHandleStyles(datum: CalloutProperties, handle: 'start' | 'end') {
+        return handle === 'start' ? super.getHandleStyles(datum, handle) : { fill: undefined, strokeWidth: 0 };
+    }
+
+    protected override updateAnchor(datum: CalloutProperties, bbox: _Scene.BBox, context: AnnotationContext) {
+        const coords = convertLine(datum, context);
+        if (!coords) {
+            return this.anchor;
+        }
+        const { bodyBounds } = this.getDimensions(datum, bbox, coords) ?? {};
+        const bounds = bodyBounds ?? bbox;
+        return {
+            x: bounds.x + context.seriesRect.x,
+            y: bounds.y + context.seriesRect.y - bounds.height,
+            position: this.anchor.position,
+        };
+    }
+
+    protected override updateShape(datum: CalloutProperties, textBBox: _Scene.BBox, coords: LineCoords) {
+        const { shape } = this;
+
+        // update shape styles
+        shape.fill = datum.fill;
+        shape.fillOpacity = datum.fillOpacity ?? 1;
+        shape.stroke = datum.stroke;
+        shape.strokeWidth = datum.strokeWidth ?? 1;
+        shape.strokeOpacity = datum.strokeOpacity ?? 1;
+
+        // update shape path
+
+        const { tailPoint, bodyBounds } = this.getDimensions(datum, textBBox, coords) ?? {};
+        if (!tailPoint || !bodyBounds) {
+            return;
+        }
+
+        this.updatePath(tailPoint, bodyBounds);
+    }
+
+    private updatePath(tailPoint: Coords, bodyBounds: Bounds) {
+        const { x: tailX, y: tailY } = tailPoint;
+        const { x, y, width, height } = bodyBounds;
+
+        const top = y - height;
+        const right = x + width;
+
+        const placement = this.calculateCalloutPlacement({ x: tailX, y: tailY }, bodyBounds);
+        const cornerRadius = Math.min(12, Math.min(width, height) / 2);
+
+        const { path } = this.shape;
+        path.clear();
+
+        path.moveTo(x, top + cornerRadius);
+        this.drawPath(
```

Comment:
nit: I think a cleaner approach here would be to create the basic points first as an array then iterate through them, calling small draw functions for each side and corner. Then draw the "callout" version of the side or corner based on the index and placement.

Something like:

```ts
const points = [
    { x0: ... },
    // ...
]

for (const [index, point] of points.entries()) {
    if (index === 0 && placement === 'topLeft') {
        this.drawCalloutCorner(path, {
            ...point,
            cx: tailX,
            cy: tailY,
        }, cornerRadius);
    } else if (index === 1 && placement === 'top') {
        this.drawCalloutSide(path, point, cornerRadius);
    // else etc...
    } else if (index % 2 === 0) {
        this.drawCorner(path, point, cornerRadius);
    } else {
        this.drawSide(path, point);
    }
}
```

Thoughts?

There may be a better final approach but in principle:
- iterating rather than sequentially calling the same function multiple times
- having small functions for each case rather than a long switch statement